core/trino-main/src/main/java/io/trino/operator/exchange/ScaleWriterPartitioningExchanger.java

```diff
@@ -0,0 +1,154 @@
+        this.partitionFunction = requireNonNull(partitionFunction, "partitionFunction is null");
+        this.partitionRebalancer = requireNonNull(partitionRebalancer, "partitionRebalancer is null");
+        this.onPageReleased = forLocalExchangeMemoryManager(memoryManager);
+
+        // Initialize writerAssignments with the buffer size
+        writerAssignments = new IntArrayList[buffers.size()];
+        for (int i = 0; i < writerAssignments.length; i++) {
+            writerAssignments[i] = new IntArrayList();
+        }
+
+        // Initialize partitionAssignments with the partition limit
+        partitionAssignments = new IntArrayList[partitionCount];
+        for (int i = 0; i < partitionAssignments.length; i++) {
+            partitionAssignments[i] = new IntArrayList();
+        }
+    }
+
+    @Override
+    public void accept(Page page)
+    {
+        Page partitionPage = partitionedPagePreparer.apply(page);
+
+        // Assign each row to a partition which limits to the partitionCount. If there are more physical partitions than
+        // this artificial partition limit, then it is possible that multiple partitions will get assigned the same
+        // bucket id. Thus, multiple partitions will be scaled together since we track partition physicalWrittenBytes
+        // using the artificial limit (partitionCount). The assignments lists are all expected to cleared by the
+        // previous iterations.
+        for (int position = 0; position < partitionPage.getPositionCount(); position++) {
+            int partition = partitionFunction.getPartition(partitionPage, position);
+            partitionAssignments[partition].add(position);
+        }
+
+        // Scale up writers when current buffer memory utilization is more than 50% of the maximum
+        if (buffers.size() > 1 && memoryManager.getBufferedBytes() > maxBufferedBytes * 0.5) {
+            partitionRebalancer.rebalancePartitions();
+        }
+
+        // assign each partition to a specific writer
+        for (int partitionId = 0; partitionId < partitionAssignments.length; partitionId++) {
+            IntArrayList positionsList = partitionAssignments[partitionId];
+            int partitionSize = positionsList.size();
+            if (partitionSize == 0) {
+                continue;
+            }
+
+            // clear the assigned positions list size for the next iteration to start empty. This
+            // only resets the size() to 0 which controls the index where subsequent calls to add()
+            // will store new values, but does not modify the positions array
+            int[] positions = positionsList.elements();
+            positionsList.clear();
+
+            int nextPartitionWriterId = partitionRebalancer.getNextWriterId(partitionId);
+            partitionRebalancer.addPartitionRowCount(nextPartitionWriterId, partitionId, partitionSize);
```

Comment:
> this may end up being called on every single row.

You mean in the worst case right? Because we are calling this for every partition present in a page

Reply:
Yeah